vtex/faststore
# Refactor |  Search Context

## What's the purpose of this pull request?
This PR decreases search complexity by moving search state management to `@vtex/store-sdk` and unifying some APIs into a search context. Also, we split `search.tsx` template into `search.server.tsx` and `search.browser.tsx` so we get closer to using the Gatsby's file system API. Also, this PR fixes the Gatsby's onInitClientRender callback and moves hydration flag in there so we have a better UX when navigating client-side

## How it works?
The store-sdk package defines a standard for creating and modifying faceted search state. This state can be serialized into a `URL` object so UI developers can use the browser's History API to actively store their faceted search state. This way, one can create a rich, router-independent way for applying and removing facets. 

This differs from a previous solution by serializing all search params into the URL's path. This is nice for Jamstack because our servers can better chose the right template for a given search. For instance, this is a search page filter by descending price before and after this PR 
```
/cozinha?orderBy=OrderByPriceDESC&map=c
/cozinha/price-desc?map=c%2Csort
```
as you can see, the `?orderBy=OrderByPriceDESC` querystring became a pair of `price-desc` in path and `sort` in the `map` querystring. With this feature, we can know pre-render search pages with a custom sorting criteria. We couldn't do this before because the same page would be responded by our server regardless of the querystring `orderBy`.

Also, this PR improves the search context provider in theme-store by adding some functionalities for easily toggling and setting multiple facets. Setting multiple facets will be very helpful for creating a decent mobile experience where the user choses multiple different facets and them apply the changes.

## How to test it?
Things you should test:
1. In a search page, toggle multiple facet filters. Each filter you select must appear on the browser's path
2. In a search page, change the priceRange facet. There should only be one intance of the parameter on the path and it should reflect what you see on the UI
3. Go to boti's `/sale`. This landing page should work normally and all filters should be set correctly
4. In a search page, change the sort criteria of the search. This should be reflected on the browser's path and should be only one instance of this filter
5. In a regionalized page, you should always see a `/per` in the end of the url, meaning you are in a regionalized search page
6. Check that all meta tags are identical from what's on production today to what's going to production

https://github.com/vtex-sites/btglobal.store/pull/534
https://github.com/vtex-sites/marinbrasil.store/pull/500
https://github.com/vtex-sites/storecomponents.store/pull/896
https://github.com/vtex-sites/carrefourbrfood.store/pull/636

## Future work?
This PR opens a way to:
1. Add SearchPage to views
2. Use FileSystem routing API

Needs improvement
1. I'm not happy with regionalized search. I'm just copying today's default behavior, but I fell this should be changed
